### PR TITLE
feat: CLI option to use ParityDB and disable RocksDB to build Substrate without pulling its dependencies

### PR DESCRIPTION
**Pull request by ltfschoen**
- [ ] **Description:** You added a brief description of the PR, e.g.:
  - What does it do?

Whilst currently `ParityDB` can be enabled with a feature flag, and is still experimental, doing so will still pull all the `RocksDB` dependencies. So this PR is to update the CLI to include more flags during `cargo build`, to give "brave" users the choice of being able to disable the pulling of RocksDB dependencies if they wish to only use ParityDB. This might help some users run Substrate on Android if they want to, even though it's not officially supported (see open issues mentioned below).

This is a WIP. Currently focussing on Phase 1.

* [X] Phase 1: Disable RocksDB functionality entirely (comment it out) and replace with only ParityDB and compiles successfully and runs with ParityDB by default when you run `cargo build && ./target/debug/substrate --dev --base-path=/tmp/data --rpc-external --ws-external --rpc-cors=all --pruning=archive`

* Phase 2: Update CLI and [`[features]`](https://doc.rust-lang.org/cargo/reference/features.html) or similar to conditionally `cargo build` using only ParityDB without fetching any RocksDB dependencies but being able to restore the RocksDB dependencies in the code so they have a choice.
* Phase 3: Update weights in all Frame's to be conditionally applied depending on what weights the user requests from the CLI, since at the moment they are all using only `constants::RocksDbWeight` rather than `ParityDbWeight`.  
  * Note: It states that `ParityDB` can be enabled with a feature flag, but is still experimental. These weights
/// are available for brave runtime engineers who may want to try this out as default.
{{else}}
/// By default, Substrate uses `RocksDB`, so this will be the weight used throughout
/// the runtime.

[Polkachu]: https://polkachu.com/snapshots
[paritydb_weights.rs]: https://github.com/paritytech/polkadot/blob/c254e5975711a6497af256f6831e9a6c752d28f5/runtime/polkadot/constants/src/weights/paritydb_weights.rs#L60

* Phase 4: Update CLI 

  - What important points should reviewers know?

  - Is there something left for follow-up PRs?

- [ ] **Related Issues:** You mentioned a related issue if this PR is related to it, e.g. `Fixes #228` or `Related #1337`.
* https://github.com/paritytech/substrate/issues/13354
* https://github.com/paritytech/substrate/issues/13177

- [ ] **2 Reviewers:** You asked at least two reviewers to review. If you aren't sure, start with GH suggestions.
* TODO - when its ready for review
- [X] **Style Guide:** Your PR adheres to [the style guide](https://github.com/paritytech/substrate/blob/master/docs/STYLE_GUIDE.md)
  - In particular, mind the maximal line length of 100 (120 in exceptional circumstances).
  - There is no commented code checked in unless necessary.
  - Any panickers in the runtime have a proof or were removed.
- [ ] **Runtime Version:** You bumped the runtime version if there are breaking changes in the **runtime**.
* TODO - if necessary
- [ ] **Docs:** You updated any rustdocs which may need to change.
* TODO - if necessary
- [ ] **Polkadot Companion:** Has the PR altered the external API or interfaces used by Polkadot?
  - [ ] If so, do you have the corresponding Polkadot PR ready?
  - [ ] Optionally: Do you have a corresponding Cumulus PR?
  * TODO - if necessary

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512090

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512077

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512079

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512074

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512078

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512084

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512089

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512088

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512073

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512208

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512195

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512197

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512206

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512192

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512202

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512207

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512196

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512191

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512267

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2512316

File: utils/frame/benchmarking-cli/Cargo.toml [hunk 1]
```diff
@@ -55,6 +55,6 @@ sp-trie = { version = "7.0.0", path = "../../../primitives/trie" }
 gethostname = "0.2.3"
 
 [features]
-default = ["rocksdb"]
+default = ["paritydb"]
```
Comment (bkchr) [inline utils/frame/benchmarking-cli/Cargo.toml:58]: Rocksdb is already an optional feature. Not sure why you are then changing this. This means that you could disable rocksdb already.

Comment (ltfschoen) [reply utils/frame/benchmarking-cli/Cargo.toml:58]: thanks, i'm brushing up on the Rust documentation :)

**Comment by ltfschoen** [comment]
@bkchr i was stuck on what to do with the `librocksdb-sys = { opt-level = 3 }` dependency that's in the workspace manifest file.
i couldn't add the following in the virtual manifest workspace Cargo.toml file
```
[workspace.dependencies]
librocksdb-sys = { optional = true }
```
because it gave error `virtual manifest specifies a [dependencies] section, which is not allowed`
and i could't do the following:
```
[features]
use-librocksdb-sys = ["dep:librocksdb-sys", "kvdb-rocksdb?/librocksdb-sys"]
```
because it gave error `virtual manifest specifies a [features] section, which is not allowed`

and i could't do the following in client/db/Cargo.toml
```
[dependencies]
...
kvdb-rocksdb = { version = "0.17.0", optional = true }
librocksdb-sys = { optional = true, opt-level = 3 }

...

[features]
use-librocksdb-sys = [
	"dep:librocksdb-sys",
	"dep:kvdb-rocksdb",
]
```
because it gave error `unused manifest key: dependencies.librocksdb-sys.opt-level`.

so i tried just removing all dependency imports of `kvdb-rocksdb` (since `librocksdb-sys` is a dependency of it), and i only left `librocksdb-sys = { opt-level = 3 }` in the workspace Cargo.toml file, and when i ran `cargo build` it gave warning `warning: profile package spec librocksdb-sys in profile 'dev' did not match any packages`, but the Cargo.lock file that it generated didn't contain `librocksdb-sys` so hadn't pulled the RocksDB dependencies, which means i can leave that workspace Cargo.toml file alone.
 
so then i found that i'll have to remove `kvdb-rocksdb from `[dev-dependencies]` in client/db/Cargo.toml because you can't use `optional = true` there to be able to enable or disable the feature (i.e. choose to use paritydb or rocksdb), you can only use `optional = true` in `[dependencies]`.

so since `librocksdb-sys` `0.8.3+7.4.4` is a dependency of `kvdb-rocksdb` `0.17.0`, i think i just need to enable a feature to enable the use of the dependency `kvdb-rocksdb` in packages that use it.

so then in client/db/Cargo.toml i added `kvdb-rocksdb` but commented out:
```
[dependencies]
# kvdb-rocksdb = { version = "0.17.0", default-features = false, optional = true }
...
[features]
# kvdb-rocksdb = ["dep:kvdb-rocksdb"]
```
and running `cargo build` created a Cargo.lock file, that didn't contain `librockdb-sys`, which was expected.

then i uncommented those lines
```
[dependencies]
kvdb-rocksdb = { version = "0.17.0", default-features = false, optional = true }
...
[features]
default = []
kvdb-rocksdb = ["dep:kvdb-rocksdb"]
```

but i don't understand why the Cargo.lock now contained 
```
[[package]]
name = "librocksdb-sys"
version = "0.8.3+7.4.4"
```
because i didn't pass in `kvdb-rocksdb` as a feature option to the build command like `cargo build --feature kvdb-rocksdb`, and i didn't have any default features, i had only run `cargo build`.

**Comment by ltfschoen** [comment]
maybe a better approach too would be to have a platform-specific build-dependencies section that is dedicated to `aarch64-linux-android` https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#build-dependencies
```
[target.'cfg(not(target_os = "android"))'.build-dependencies]
kvdb-rocksdb = "0.17.0"
```
so on android you'd build with `cargo build --target=aarch64-linux-android`

**Comment by juangirini** [comment]
@ltfschoen @bkchr should this be reopen?

**Pull request closed without merging**